crates/node/src/networking/p2p/protocol/mod.rs

```diff
@@ -0,0 +1,186 @@
+use eyre::{eyre, Result};
+
+pub mod internal;
+pub mod v0;
+pub mod v1;
+
+pub fn serialize_handshake(msg: internal::Handshake) -> Result<Vec<u8>> {
+    let msg = v0::Handshake::from(msg);
+    bincode::serialize(&msg).map_err(|e| e.into())
+}
+
+pub fn new_serialize_handshake(protocol_version: u64, msg: internal::Handshake) -> Result<Vec<u8>> {
+    match protocol_version {
+        0 => bincode::serialize(&v0::Handshake::from(msg)),
+        1 => bincode::serialize(&v1::Handshake::from(msg)),
+        ver => return Err(eyre!("unknown protocol version: {ver}")),
+    }
+    .map(|mut bs| {
+        let mut data = Vec::with_capacity(bs.len() + 8);
+        data.append(&mut protocol_version.to_be_bytes().to_vec());
+        data.append(&mut bs);
+        data
+    })
+    .map_err(|e| e.into())
+}
+
+pub fn deserialize_handshake(bs: &[u8]) -> Result<internal::Handshake> {
+    let res = v0::Handshake::parse(bs);
+    if res.is_ok() {
+        return res;
+    }
+
+    if bs.len() < 9 {
+        return Err(eyre!("invalid handshake message: too short"));
+    }
+
+    match u64::from_be_bytes(bs[0..8].try_into().expect("convert slice to array")) {
+        0 => v0::Handshake::parse(&bs[8..]),
+        1 => v1::Handshake::parse(&bs[8..]),
+        ver => Err(eyre!("unknown protocol version: {ver}")),
+    }
+}
```

Comment:
`deserialize_handshake` is backwards compatible. First it tries the present state where individual messages are not prepended with version and only then it tries versioned approach.